src/main/java/org/zalando/nakadi/service/CursorOperationsService.java

```diff
@@ -0,0 +1,169 @@
+            return new NakadiCursorDistanceResult(query, distance);
+        } else if (initialCursor.getTimeline().getOrder() > finalCursor.getTimeline().getOrder()) {
+            throw new InvalidCursorOperation(INVERTED_TIMELINE_ORDER);
+        } else {
+            final long distance = getDistanceDifferentTimelines(query, initialCursor, finalCursor);
+            return new NakadiCursorDistanceResult(query, distance);
+        }
+    }
+
+    public NakadiCursor unshiftCursor(final ShiftedNakadiCursor cursor) throws InvalidCursorOperation {
+        if (cursor.getShift() < 0) {
+            return moveCursorBackwards(cursor.getTopic(), cursor.getTimeline(), cursor.getPartition(),
+                    parse(cursor.getOffset()), cursor.getShift());
+        } else {
+            return moveCursorForward(cursor.getTopic(), cursor.getTimeline(), cursor.getPartition(),
+                    parse(cursor.getOffset()), cursor.getShift());
+        }
+    }
+
+    private NakadiCursor moveCursorBackwards(final String topic, final Timeline timeline, final String partition,
+                                             final long offset, final long shift) {
+        final long shiftedOffset = offset + shift;
+        if (shiftedOffset >= 0) { // move left in the same timeline
+            final String paddedOffset = getPaddedOffset(shiftedOffset);
+            return new NakadiCursor(timeline, partition, paddedOffset);
+        } else { // move to previous timeline
+            final Timeline previousTimeline = getTimeline(topic, timeline.getOrder() - 1);
+            final long totalEventsInTimeline = totalEventsInTimeline(previousTimeline, partition);
+            return moveCursorBackwards(topic, previousTimeline, partition, totalEventsInTimeline, shift + offset);
+        }
+    }
+
+    private NakadiCursor moveCursorForward(final String topic, final Timeline timeline, final String partition,
+                                    final long offset, final long shift) {
+        if (timeline.isClosed()) {
+            if (offset + shift < totalEventsInTimeline(timeline, partition)) { // move right in the same timeline
+                final long finalOffset = offset + shift;
+                final String paddedOffset = getPaddedOffset(finalOffset);
+                return new NakadiCursor(timeline, partition, paddedOffset);
+            } else { // move right, next timeline
+                final Timeline nextTimeline = getTimeline(topic, timeline.getOrder() + 1);
+                final long totalEventsInTimeline = totalEventsInTimeline(nextTimeline, partition);
+                final long newShift = shift - (totalEventsInTimeline - offset);
+                return moveCursorForward(topic, nextTimeline, partition, 0, newShift);
+            }
+        } else { // Open timeline. Avoid checking for the latest available offset for performance reasons
```

Comment:
I don't know. For me it's fine as it is. In the worst case it will be the cursor that points so something that doesn't exist yet. Anyway when user will try to read from it - he will get the error.